chain/gap/find.go

```diff
@@ -0,0 +1,147 @@
+package gap
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/sentinel-visor/chain"
+	"github.com/filecoin-project/sentinel-visor/lens"
+	"github.com/filecoin-project/sentinel-visor/model/visor"
+	"github.com/filecoin-project/sentinel-visor/storage"
+	"golang.org/x/xerrors"
+)
+
+type GapIndexer struct {
+	DB     *storage.Database
+	opener lens.APIOpener
+}
+
+func NewGapIndexer(o lens.APIOpener, db *storage.Database) *GapIndexer {
+	return &GapIndexer{
+		DB:     db,
+		opener: o,
+	}
+}
+
+func (g *GapIndexer) Run(ctx context.Context) error {
+	node, closer, err := g.opener.Open(ctx)
+	if err != nil {
+		return xerrors.Errorf("open lens: %w", err)
+	}
+	defer func() {
+		closer()
+	}()
+
+	findLog := log.With("type", "find")
+	heightGaps, err := g.detectProcessingGaps(ctx, node)
+	if err != nil {
+		return xerrors.Errorf("detecting processing gaps: %w", err)
+	}
+
+	findLog.Infow("detected gaps in height", "count", len(heightGaps))
+
+	skipGaps, err := g.detectSkippedTipSets(ctx, node)
+	if err != nil {
+		return xerrors.Errorf("detecting skipped gaps: %w", err)
+	}
+
+	findLog.Infow("detected gaps from skip", "count", len(skipGaps))
+
+	return g.DB.PersistBatch(ctx, skipGaps, heightGaps)
+}
+
+func (g *GapIndexer) detectSkippedTipSets(ctx context.Context, node lens.API) (visor.GapReportList, error) {
+	reportTime := time.Now()
+	// TODO unsure how big these lists will be, will likely want some sort of pagination here with limits and a loop over it
+	var skippedReports []visor.ProcessingReport
+	if err := g.DB.AsORM().Model(&skippedReports).Order("height desc").Where("status = ?", visor.ProcessingStatusSkip).Select(); err != nil {
+		return nil, xerrors.Errorf("query processing report skips: %w", err)
+	}
+	gapReport := make([]*visor.GapReport, len(skippedReports))
+	for idx, r := range skippedReports {
+		tsgap, err := node.ChainGetTipSetByHeight(ctx, abi.ChainEpoch(r.Height), types.EmptyTSK)
+		if err != nil {
+			return nil, xerrors.Errorf("getting tipset by height %d: %w", r.Height, err)
+		}
+		gapReport[idx] = &visor.GapReport{
+			Height:     r.Height,
+			TipSet:     tsgap.Key().String(),
+			Task:       r.Task,
+			Status:     "GAP",
+			Reporter:   "gapIndexer", // TODO does this really need a name?
+			ReportedAt: reportTime,
+		}
+	}
+	return gapReport, nil
+}
+
+func (g *GapIndexer) detectProcessingGaps(ctx context.Context, node lens.API) (visor.GapReportList, error) {
+	reportTime := time.Now()
+	// TODO unsure how big these lists will be, will likely want some sort of pagination here with limits and a loop over it
+
+	// build list of all reports ordered by height, scan these to find gaps in epochs
+	var reportHeights []visor.ProcessingReport
+	if err := g.DB.AsORM().Model(&reportHeights).
```

Comment:
This is a lot of data. Soon to be a million rows.

Reply:
seems like you could select reports where an adjacently numbered report doesnt exist.

Reply:
I think a lot could be improved here:
1. Configurable pagination based on height.
2. Clever SQL, e.g. `select * from visor_processing_reports where task=<task> and (select count(1) from visor_processing_reports where height = rep.height + 1) = 0;` 

will adjust.